tests: repro for zombie txs after compaction

andrwng opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

    This reproduces a bug fixed by e8e9ab7f27056d31b5287b2f6bc6cccae0e703fa
    wherein an aborted transaction was visible by a consumer after
    compaction and replication.

    The bug was caused by the fact that we would skip over segments of size
    1, rather than self compacting them.

    The sequence is roughly:

    node 1:
    - begin tx 1 (seg 1)
    - roll segment
    - data batch in tx 1 (seg 2)
    - roll segment
    - abort tx 1 (seg 3)
    - begin tx 2, data batch in tx 2, abort tx 2 (seg 3)
    - run compaction: skips first two segments from tx 1, removes abort
      marker tx 1 because it's deduplicated in favor of abort tx 2's abort
      marker (all aborts markers have the same key)
    - replicate to node 2

    node 2:
    - gets everything remaining in a single segment, now including just
      [begin tx 1, data tx 1, abort tx 2]
    - self compaction runs, but tx 1 is not an ongoing aborted transaction
    - data tx 1 is left in the segment as committed data
 
Along the way, I made a few changes to `tx_compaction_utils` to be reused in this test.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.2.x
- [ ] v23.1.x
- [ ] v22.3.x

## Release Notes

* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bharathv reviewed (approved):
this is pretty cool.

andrwng commented:
Force push:
- rebased on recent changes to tx_compaction_utils to run housekeeping during execution
- added [change](https://github.com/redpanda-data/redpanda/pull/15672/commits/c24288e8b80dc04ad126184d8fd4f6e1ac95d4e0) to avoid running validation if not compacted (used in some gtests, but was hidden before because we were only using BOOST macros before)

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/42891#018c6ece-38ea-4187-bfab-30a536e68826
<!--53d53ce8-82d3-43e0-9267-83f508a644ca-->
